### PR TITLE
Implement GitSourceControl

### PR DESCRIPTION
**Pull request by saphal1998**
# Summary

- Filled the stubs in SourceControl for GitSourceControl except 
- Marked with Comments to see logic flow (Please let me know if something's unclear)

### CODE REVIEW

File: src/source-control/GitSourceControl.ts [hunk 1]
```diff
@@ -1,42 +1,396 @@
...
+      queue.push(
+        ...commit.childCommits.map((hash) => this.commitHashMap.get(hash)!),
+      );
+    }
+    // SEE: Call Load Repository Information
   }
 
   async attachSttackBranchesToCommits(commits: Commit[]): Promise<void> {
-    // TODO: Implement
+    // TODO: Unsure
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: We can use @manyaagarwal's implementation in #30 here, but we need to undo this change to prevent merge conflicts

```suggestion
    // TODO: Implement
```

File: src/source-control/GitSourceControl.ts [hunk 2]
```diff
@@ -1,42 +1,396 @@
...
+import fs from "fs";
+enableMapSet();
+
+import type { Commit, CommitSignature, Repository } from "../shared/types";
 import type {
   SourceControl,
   SourceControlRepositoryUpdateListener,
 } from "./SourceControl";
 
+const localRefPrefix = "refs/heads/";
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: @manyaagarwal looks like you don't have to port these functions over from #30

**Review by taneliang — COMMENTED** [review]
Nice! You may want to coordinate with @manyaagarwal to see which PR should be landed first, since both this and #30 will conflict (and have overlaps)

File: src/source-control/GitSourceControl.ts (hunk 1 above)
Comment (manyaagarwal) [reply src/source-control/GitSourceControl.ts]: Ya, I have shifted the branch name code here. The branch creation would still have to be implemented. Pushing my changes soon. Shifting stuff around.

Comment (taneliang) [reply src/source-control/GitSourceControl.ts]: Okay #30 doesn't implement this. We'll need to generate branch names (which #30 does), then do the `git branch -f` equivalent that our existing code is already doing. I can find this out for you if you need; we already have the code to do this.

#30 will need a different method signature though, so you'll need to sync with @manyaagarwal

**Comment by saphal1998** [comment]
I have some doubts on the attachSttackAttackCommits function, basically I'm asking what it's purpose is. A little confused on that. Can't we just do that in the other functions?

File: src/shared/types.ts [hunk 3]
```diff
@@ -50,5 +50,5 @@ export interface Repository {
    * This is not a `Map` as we need to be able to detect that this object has
    * changed. We could revisit this decision again in the future.
    */
-  commits: { [hash: string]: Commit };
+  commits: Map<string, Commit>;
```
Comment (manyaagarwal) [inline src/shared/types.ts]: We could make a CommitHash type like BranchName maybe?

File: src/source-control/SourceControl.ts [hunk 4]
```diff
@@ -43,7 +48,7 @@ export interface SourceControl {
    * @param targetCommit The commit the tree should be rebased on.
    * @returns Promise that resolves when operation is complete.
    */
-  rebaseCommits(rebaseRootCommit: Commit, targetCommit: Commit): Promise<void>;
+  rebaseCommits(rebaseRootCommit: string, targetCommit: string): Promise<void>;
```
Comment (manyaagarwal) [inline src/source-control/SourceControl.ts]: very small nit: I think we should be using `CommitHash` here too instead of `string`

File: src/shared/types.ts [hunk 5]
```diff
@@ -50,5 +50,5 @@ export interface Repository {
    * This is not a `Map` as we need to be able to detect that this object has
```
Comment (taneliang) [inline src/shared/types.ts]: This comment needs to be removed

File: src/source-control/GitSourceControl.ts [hunk 6]
```diff
@@ -1,47 +1,449 @@
...
   repositoryUpdateListener: SourceControlRepositoryUpdateListener;
+  private commitHashMap = new Map<string, Commit>();
 
   constructor(
     repoPath: string,
     repositoryUpdateListener: SourceControlRepositoryUpdateListener,
   ) {
     this.repoPath = repoPath;
     this.repositoryUpdateListener = repositoryUpdateListener;
+    this.populateGitSourceControl();
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Oh this is pretty dangerous, as you may end up trying to use the data before it’s ready. Could we avoid this?

File: src/source-control/GitSourceControl.ts [hunk 7]
```diff
@@ -1,47 +1,449 @@
...
+
+    this.commitHashMap = produce(this.commitHashMap, (draftCommitHashMap) => {
+      refs.forEach((ref) => {
+        const branchName = ref.name();
+        const branchSha = ref.target().tostrS();
+        draftCommitHashMap.get(branchSha)?.refNames.push(branchName);
+      });
+    });
+
+    // Inject branch names
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts:196]: Looks like this comment needs to be moved up

File: src/source-control/GitSourceControl.ts [hunk 8]
```diff
@@ -1,47 +1,449 @@
...
+      const updatedRepository = {
+        ...ourRepository,
+        commits: remoteUpdatedCommitHashMap,
+      };
+      return updatedRepository;
+    };
+
+    return {
+      repo: ourRepository,
+      remoteRepoInfoPromise: getRemoteRepoInfo(ourRepository),
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: We no longer need this promise since it’s now Stacker’s job to get PR info (cc @manyaagarwal)

File: src/source-control/SourceControl.ts (hunk 4 above)
Comment (taneliang) [inline src/source-control/SourceControl.ts]: Why should this be changed to strings? I thought you can get more info from the commit objects

File: src/source-control/GitSourceControl.ts (hunk 8 above)
Comment (manyaagarwal) [reply src/source-control/GitSourceControl.ts]: @taneliang Do you mean remote repo information?

File: yarn.lock [hunk 9]
```diff
@@ -5759,6 +5774,14 @@ ramda@^0.25.0:
   resolved "https://registry.yarnpkg.com/ramda/-/ramda-0.25.0.tgz#8fdf68231cffa90bc2f9460390a0cb74a29b29a9"
   integrity sha512-GXpfrYVPwx3K7RQ6aYT8KPS8XViSXUVJT1ONhoKPE9VAleW42YE+U+8VEyGWt41EnEQW7gwecYJriTI0pKoecQ==
 
+random-word@^2.0.0:
```
Comment (manyaagarwal) [inline yarn.lock]: These packages should be removed!

File: src/source-control/SourceControl.ts (hunk 4 above)
Comment (saphal1998) [reply src/source-control/SourceControl.ts]: Yeah, it's an either/or tbh

File: src/source-control/GitSourceControl.ts (hunk 6 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: How do you suggest I go about that?

File: src/source-control/SourceControl.ts (hunk 4 above)
Comment (taneliang) [reply src/source-control/SourceControl.ts]: Actually I agree with your choice of using `string`. That'll allow you to (at some point in the future) handle command line commands more efficiently, as you'll only need to fetch only the commits that are requested, instead of preloading all the `Commit` objects for someone else to filter.

Could you also update the Stacker code that calls this, as well as the frontend code that calls it?

File: src/source-control/GitSourceControl.ts (hunk 6 above)
Comment (taneliang) [reply src/source-control/GitSourceControl.ts]: The cleanest option I can think of is to add a private function like `private async getGuaranteedPopulatedData` or similar, that either returns cached data if it exists or loads new data. You can then use that everywhere you need preloaded data.

Feel free to think of other options, but this is the best one I can think of right now

Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: Implemented an alternate, let me know if you think it's okay

File: src/source-control/GitSourceControl.ts [hunk 10]
```diff
@@ -19,15 +31,18 @@ function isSttackBranch(branch: BranchName): boolean {
   return branch.startsWith("sttack-");
```
Comment (manyaagarwal) [inline src/source-control/GitSourceControl.ts:31]: Is this a merge mistake? 2 return statements in one function

Comment (saphal1998) [reply src/source-control/GitSourceControl.ts:31]: Yes, probably, lol, lemme fix this

Comment (saphal1998) [reply src/source-control/GitSourceControl.ts:31]: Lol, wait no, you collapsed the lines in between. I think, right?

Comment (manyaagarwal) [reply src/source-control/GitSourceControl.ts:31]: Oh shit sorry, that was stupid :sweat:

File: src/cli/useInteractionReducer.ts [hunk 11]
```diff
@@ -73,7 +73,10 @@ function backendCommitGraphToDisplayCommits(
     hasFork: boolean,
   ): DisplayCommit[] {
     const childCommits = subgraphBackendRootCommit.childCommits.map(
-      (childCommitHash) => repository.commits[childCommitHash],
+      /*  SEE - Changes from Object Ref to Map.get, making the retutn type Commit | undefined. 
+      Adding a `!` here to get rid of TS errors
+      */
```
Comment (taneliang) [inline src/cli/useInteractionReducer.ts]: This makes sense, but I think it'll be better if the comment explains the code rather than the change (because once this PR is merged the comment won't make sense)
```suggestion
      // Since the commits came from the graph, childCommitHash must exist in repository.commits.
```

File: src/source-control/SourceControl.ts [hunk 12]
```diff
@@ -66,4 +69,10 @@ export interface SourceControl {
   attachSttackBranchesToCommits(
     commits: Commit[],
   ): Promise<Array<{ commit: Commit; sttackBranch: BranchName }>>;
+
+  /**
+   * Getter for Repo from Source Control
+   * @returns Promise that resolves into the repository
+   */
+  getRepo(): Promise<Repository>;
```
Comment (taneliang) [inline src/source-control/SourceControl.ts]: What's this for? Won't we want to use the listener instead of a promise like this?

File: src/stacker/Stacker.ts [hunk 13]
```diff
@@ -30,11 +30,14 @@ export interface Stacker {
    *
    * Calls `repositoryUpdateListener` on success.
    *
-   * @param rebaseRootCommit The commit tree root to be moved.
-   * @param targetCommit The commit the tree should be rebased on.
+   * @param rebaseRootCommit The commit has for tree root to be moved.
```
Comment (taneliang) [inline src/stacker/Stacker.ts]:
```suggestion
   * @param rebaseRootCommit The commit hash for tree root to be moved.
```

File: src/source-control/SourceControl.ts (hunk 12 above)
Comment (saphal1998) [reply src/source-control/SourceControl.ts]: Yeah, as discussed on Discord, this function is kind of redundant. Removing this.

**Comment by saphal1998** [comment]
All Resolved! 💯 

Should I merge this then?  @taneliang @manyaagarwal @jessieAnhNguyen

File: src/source-control/GitSourceControl.ts [hunk 14]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+    const repo = await nodegit.Repository.open(this.repoPath);
+    const repoStatus = await repo.getStatus();
+    if (repoStatus.length === 0) {
+      // No changes in repo, can return
+      return;
+    } else {
+      await this.populateGitSourceControl();
+      this.loadRepositoryInformation();
+    }
+    // Else we can call populate GitSourceControl
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Does this describe the else block above, or is this a TODO for future code?

File: src/source-control/GitSourceControl.ts [hunk 15]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+        return;
+      }
+      childrenHashes.forEach((childHash) => {
+        const childCommit: Commit = hashMap.get(childHash)!;
+        childCommit.parentCommits = Array.from(
+          new Set([...childCommit.parentCommits, parentHash]),
+        );
+        parentCommit.childCommits = Array.from(
+          new Set([...parentCommit.childCommits, childHash]),
+        );
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Nit but I think it'll be clearer (and slightly more performant and also results in a deterministic order) if we just check if the hash exists in the array instead of constructing a new set

```suggestion
        if (!childCommit.parentCommits.includes(parentHash)) {
          childCommit.parentCommits = [...childCommit.parentCommits, parentHash];
        }
        if (!parentCommit.childCommits.includes(parentHash)) {
          parentCommit.childCommits = [...parentCommit.childCommits, childHash];
        }
```

File: src/source-control/GitSourceControl.ts [hunk 16]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+      });
+      hashMap.set(parentHash, parentCommit);
+    });
+
+    // Inject branch names
+
+    refs.forEach((ref) => {
+      const branchName = ref.name();
+      const branchSha = ref.target().tostrS();
+      hashMap.get(branchSha)?.refNames.push(branchName);
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Probably a relic from the old code, but `branchName` isn't accurate
```suggestion
      const refName = ref.name();
      const branchSha = ref.target().tostrS();
      hashMap.get(branchSha)?.refNames.push(refName);
```

File: src/source-control/GitSourceControl.ts [hunk 17]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+    }
+    // Else we can call populate GitSourceControl
+  }
+
+  private async populateGitSourceControl(): Promise<void> {
+    const repo = await nodegit.Repository.open(this.repoPath);
+    const repoStatus = await repo.getStatus();
+    const refs = await repo.getReferences();
+    const headHash = (await repo.getHeadCommit()).sha();
+    const hashMap = new Map<CommitHash, Commit>();
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: This is a really generic name 😆 Could we rename it to something like `commitHashMap` or `mapOfCommitHashesToCommits` or something similar? The name "hash" is pretty unfortunate since it has 2 meanings in our code (hashmap and commit hash)

File: src/source-control/GitSourceControl.ts [hunk 18]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+    const tempBranchName = `sttack-${lorem.slug(3)}`;
+    const queue: string[] = [rebaseRootCommit];
+    // 1. Rebase root commit on target commit
+    // 2. Update Target Commit to point to the cherry picked commit
+    // 3. Update rebaseRootCommit to point to child of itself
+    // Call LoadRepositoryInformation
+
+    while (queue.length) {
+      const hashToBeRebased = queue.pop();
+      // This rebase commit must exist
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: This commit is a bit ambiguous - is this a prerequisite, an assumption, or a declaration of some fact?

File: src/source-control/GitSourceControl.ts [hunk 19]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+              callbacks: {
+                credentials: function (_: string, userName: string) {
+                  return nodegit.Cred.sshKeyNew(
+                    userName,
+                    userPublicKeyPath, //"/Users/phuonganh/.ssh/id_rsa.pub",
+                    userPrivateKeyPath, //"/Users/phuonganh/.ssh/id_rsa",
+                    userPassphrase, //"hello",
+                  );
+                },
+              },
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Could we reuse `callback` from above?

```suggestion
              callbacks: callback,
```

File: src/source-control/GitSourceControl.ts [hunk 20]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+    } catch (err) {
+      console.log(err);
+    }
   }
 
   async pushCommitsForCommitTreeRootedAtCommit(commit: Commit): Promise<void> {
-    // TODO: Implement
+    const queue: Commit[] = [commit];
+    while (queue.length) {
+      // SEE: Always have an element there
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: What does this mean?

File: src/stacker/Stacker.ts [hunk 21]
```diff
@@ -30,11 +30,14 @@ export interface Stacker {
    *
    * Calls `repositoryUpdateListener` on success.
    *
-   * @param rebaseRootCommit The commit tree root to be moved.
-   * @param targetCommit The commit the tree should be rebased on.
+   * @param rebaseRootCommit The commit hash has for tree root to be moved.
```
Comment (taneliang) [inline src/stacker/Stacker.ts]: typo
```suggestion
   * @param rebaseRootCommit The commit hash of the tree root to be moved.
```

File: src/source-control/GitSourceControl.ts [hunk 22]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+        parentCommit.childCommits = Array.from(
+          new Set([...parentCommit.childCommits, childHash]),
+        );
+        hashMap.set(childHash, childCommit);
+      });
+      hashMap.set(parentHash, parentCommit);
+    });
+
+    // Inject branch names
+
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: nit: let's remove this empty line so that it's clearer that the comment above describes only this code block below
```suggestion
```

File: src/source-control/GitSourceControl.ts [hunk 23]
```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
...
+    );
+
+    // Find all commits and build children maps
+    /**
+     * Maps commit hashes to their childrens' commit hashes.
+     *
+     * This needs to exist as we can only retrieve a commit's parents, but we
+     * actually want to build relationships with its children.
+     *
+     * Commit hash keys *are not* guaranteed to exist in `commitHashDict`.
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: All the comments below with `commitHashDict` need to be updated

File: src/cli/useInteractionReducer.ts [hunk 24]
```diff
@@ -73,7 +73,9 @@ function backendCommitGraphToDisplayCommits(
     hasFork: boolean,
   ): DisplayCommit[] {
     const childCommits = subgraphBackendRootCommit.childCommits.map(
-      (childCommitHash) => repository.commits[childCommitHash],
+      /*  SEE - Since the commits came from the graph, childCommitHash must exist in repository.commits.
+       */
```
Comment (taneliang) [inline src/cli/useInteractionReducer.ts]: Style nits:

1. imo the "SEE" isn't necessary since this doesn't seem to be particularly important
1. I think we should prefer `//`-style comments instead of `/* */` for consistency with typical JS code style, especially if the comment is just one line

```suggestion
      // Since the commits came from the graph, childCommitHash must exist in repository.commits.
```

**Review by taneliang — APPROVED** [review]
Only nits. Hopefully the code works 😆

File: src/source-control/GitSourceControl.ts (hunk 14 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: It's for the else block above, my bad!

File: src/source-control/GitSourceControl.ts (hunk 15 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: Sounds good!

File: src/source-control/GitSourceControl.ts (hunk 16 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: 👍

File: src/source-control/GitSourceControl.ts (hunk 17 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: Named it commitHashMap 👍

File: src/source-control/GitSourceControl.ts (hunk 18 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: It's a prerequisite and also an assumption, because we populate our repo with commits prior to calling this function

Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: I'll throw an error if either of these does not exist, updating this.

File: src/source-control/GitSourceControl.ts (hunk 19 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: Sure! Will do this

File: src/source-control/GitSourceControl.ts (hunk 20 above)
Comment (saphal1998) [reply src/source-control/GitSourceControl.ts]: Yeah, one of the comments I had while writing the code. Removing this.

File: src/source-control/GitSourceControl.ts (hunk 18 above)
Comment (taneliang) [reply src/source-control/GitSourceControl.ts]: Yeah that should be okay. The frontend can prettify it eventually
